C# Improvements

eisber opened the pull request:
@lhoang29  review :)

allow models to be saved
added namespace/feature enumeration support from C# side
allow ExampleBuilder to build against example pool vs vw instance
exposing scalars (float[]) predictions
support schema specification for ADF types
support labels to be discovered by serializer
support multiclass label
allow for graceful handling of uninitialized thread pool
basic support for config files (ADF missing)
added documentation
improved nuget/msbuild target wrt zlib.dll. properly cascades zlib dependency
allow retargeting of examples from creation instance to learn/prediction instance. useful to split serialization/learning into multiple threads
added threadsafe example pool

JohnLangford commented:
@lhoang29 ping

JohnLangford commented:
Markus, it would also help to have a bit more description of what's going on here.

JohnLangford commented:
(Is this a big collection of little changes?)

eisber commented:
I put the following comment in. Are you looking for more?

@lhoang29<https://github.com/lhoang29> review :)
allow models to be saved
added namespace/feature enumeration support from C# side
allow ExampleBuilder to build against example pool vs vw instance
exposing scalars (float[]) predictions
support schema specification for ADF types
support labels to be discovered by serializer
support multiclass label
allow for graceful handling of uninitialized thread pool
basic support for config files (ADF missing)
added documentation
improved nuget/msbuild target wrt zlib.dll. properly cascades zlib dependency
allow retargeting of examples from creation instance to learn/prediction instance. useful to split serialization/learning into multiple threads
added threadsafe example pool



From: John [mailto:notifications@github.com]
Sent: Saturday, April 2, 2016 2:47 PM
To: JohnLangford/vowpal_wabbit <vowpal_wabbit@noreply.github.com>
Cc: Markus Cozowicz <marcozo@microsoft.com>
Subject: Re: [JohnLangford/vowpal_wabbit] C# Improvements (#976)


Markus, it would also help to have a bit more description of what's going on here.

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/JohnLangford/vowpal_wabbit/pull/976#issuecomment-204772854>

eisber commented:
Yes.

From: John [mailto:notifications@github.com]
Sent: Saturday, April 2, 2016 2:48 PM
To: JohnLangford/vowpal_wabbit <vowpal_wabbit@noreply.github.com>
Cc: Markus Cozowicz <marcozo@microsoft.com>
Subject: Re: [JohnLangford/vowpal_wabbit] C# Improvements (#976)


(Is this a big collection of little changes?)

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/JohnLangford/vowpal_wabbit/pull/976#issuecomment-204773199>

JohnLangford commented on vowpalwabbit/mwt.cc:
```diff
@@ -154,30 +148,68 @@ namespace MWT {
+
+    // policies
+    size_t policies_size = c.policies.size();
+    bin_text_read_write_fixed_validated(model_file, (char*)&policies_size, sizeof(policies_size), "", read, msg, text);
```
You need to update msg here, right?

eisber replied:
No, bin_text_read_write_fixed_validated sets msg to “”

JohnLangford commented on vowpalwabbit/mwt.cc:
```diff
@@ -226,6 +259,7 @@ base_learner* mwt_setup(vw& all)
     l = &init_learner(&c, setup_base(all), predict_or_learn<false, false, true>, predict_or_learn<false, false, false>, 1);
 
+  l->set_save_load(save_load);
   l->set_finish_example(finish_example);
```
Huh.  Set_finish_example _is_ set here.

JohnLangford replied:
Ignore---was confusing with cb_explore

JohnLangford commented:
@lhoang29 have you had a chance to review?

I'll reiterate: large patches are bad.  They may incorporation much more difficult because there are many more things that have to be done right.

eisber commented:
Sorry…

lhoang29 commented:
oops just saw the email, don't know why Github didn't notify me on this one. I'll take a look tomorrow

lhoang29 commented on cs/common/Reflection/ReflectionHelper.cs:
```diff
@@ -147,7 +147,7 @@ select new
                 //Debug.WriteLine("\t specializing: " + method);
             }
-            //Debug.WriteLine("Method: {0} for {1} {2}", method, name, string.Join(",", parameterTypes.Select(t => t.ToString())));
+            // Debug.WriteLine("Method: {0} for {1} {2}", method, name, string.Join(",", parameterTypes.Select(t => t.ToString())));
```
remove?

eisber replied:
This is really helpful from time to time to debug things. Maybe I should allow for tracing? Do you know about standard ways of configuring tracing for C# libs?

lhoang29 replied:
doesn't Tracing write to VW output window by default in debug mode?

eisber replied:
I'm more asking about a more general strategy for tracing in the library.

lhoang29 commented on cs/cs/Serializer/ConfigInspector.cs:
```diff
@@ -0,0 +1,211 @@
+                ",
+            RegexOptions.IgnorePatternWhitespace | RegexOptions.Compiled);
+
+        public static Schema CreateSchema(Type type, string config, Action<string> warnings = null)
```
can't seem to make out what this method is for?

eisber replied:
That’s for parsing the dynamic feature config file that Oswaldo asked for.

lhoang29 commented on cs/cs/Serializer/VowpalWabbitJsonSerializer.cs:
```diff
@@ -102,6 +102,47 @@ public void Parse(string json, ILabel label = null, int? index = null)
+
+            while (reader.Read())
+            {
+                if (!(reader.TokenType == JsonToken.PropertyName && (string)reader.Value == "_multi"))
```
don't you have this string defined somewhere?

eisber replied:
Hmm. Actually that’s the string I wanted to make configurable. What do you think? Add as settings?

lhoang29 replied:
yeah seems like a good idea

eisber replied:
implemented

lhoang29 commented on cs/cs/VowpalWabbitMultiworldTesting.cs:
```diff
@@ -0,0 +1,123 @@
+        /// Must public for the serializer to work with.
+        /// </summary>
+        [EditorBrowsableAttribute(EditorBrowsableState.Never)]
+        public sealed class LearnedVsConstantPolicy
```
hmm, not sure what I think about this. It seems a bit awkward to have a class for just the learned and constant-action policies. Would it be better to do this generally so you just have a single class with a feature vector. The caller can then decide what this vector means? (for example, in this case the first one is learned and rest is constant-action)

eisber replied:
agree

lhoang29 commented on cs/unittest/TestConfigInspector.cs:
```diff
@@ -0,0 +1,121 @@
+ns1.f2(FeatureGroup = 'a')
+ns1.f3(Namespace = ""a"",Enumerize=true, FeatureGroup = 'g'  )                
+ns1.f4(Enumerize=true , FeatureGroup ='f', Order=3,AddAnchor=true)   
+ns1.ns2.f5       
```
maybe this answers my question above. This is for ingesting the dynamic context structure?

eisber replied:
yes.
